### PR TITLE
Replace EFPs and FNTs with PNGs

### LINKED ISSUE #115: Replace .efps with .pngs
As discussed with @akheron over americano today, it would be nice to replace the bespoke .efp format with .png.

SDL_Image is an obvious contender for loading PNGs, but stb_image should also work fine. https://github.com/nothings/stb/blob/master/stb_image.h

### LINKED ISSUE #118: Replace .fnts with .pngs
Working off #115 and #117, the .fnt format should probably also be more portable.

### PR DESCRIPTION
**Pull request by akx**
This PR:

* adds Python conversion tooling for EFP -> PNG and FNT -> PNG
* replaces the Error Free Picture (EFP) assets with PNGs and replaces the loading code to look for "FOO.png" when "FOO.EFP" is requested.
* replaces the FNT assets with PNGs and replaces the loading _and_ drawing code with something somewhat saner. The original code was hard enough to make sense of. 

The game code itself (call sites to EFP loading and text drawing) remains unmodified.

Fixes #115 
Fixes #118

### CODE REVIEW

**Comment by veikkos** [comment]
Readme has libraries section and Windows specific cmake spell, which should be updated to include SDL_image.

**Comment by veikkos** [comment]
Could consider putting new files under still existing SRC/EFP directory as SRC directory has lots of files. Which reminds me that I should have actually done that with some new files introduced lately. Need to craft a PR.

File: SRC/EFP/EFP.CPP [hunk 1]
```diff
@@ -4,63 +4,21 @@
...
-        error( "load_efp(): unable to allocate %d bytes to read %s", file_len - offs, name );
-    }
-    fread( file_data, sizeof( uint8_t ), file_len - offs, efp );
-    uint8_t *data_p = file_data;
-
-    if ( strncmp((const char *) data_p, "EF pic", 6 ) != 0)
+    if ( !tk_port::load_efp_replacement_png( name, kuva, nullptr ))
     {
-        error( "File is not an EF pic!\n" );
+        error("Unable to load PNG replacement for %s", name);
```
Comment (veikkos) [inline SRC/EFP/EFP.CPP]: `\n`?

File: SRC/EFP/EFP.CPP [hunk 2]
```diff
@@ -4,63 +4,21 @@
...
 
 void load_efp_pal( const char *name, char *paletti )
 {
-    FILE *efp = fopen( name, "rb" );
-    fseek( efp, -768, SEEK_END );
-    fread( paletti, sizeof(char), 768, efp );
-    fclose( efp );
+    if ( !tk_port::load_efp_replacement_png( name, nullptr, paletti ))
+    {
+        error("Unable to load palette from PNG replacement for %s", name);
```
Comment (veikkos) [inline SRC/EFP/EFP.CPP]: `\n`?

File: SRC/PORT_IMAGE.CPP [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    bool load_png( const char *png_name, char *kuva, char *paletti, int *width, int *height )
+    {
+        SDL_RWops *rwop = SDL_RWFromFile( png_name, "rb" );
+        if ( !rwop )
+        {
+            // assume the file just didn't exist
+            return false;
+        }
+        SDL_Surface *image = IMG_LoadPNG_RW( rwop );
+        if ( !image )
```
Comment (veikkos) [inline SRC/PORT_IMAGE.CPP]: `SDL_FreeRW( rwop );` before condition to avoid leaking in case of a failure?

File: SRC/PORT_IMAGE.CPP [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                    size_t in_off = y * image->pitch;
+                    size_t out_off = y * image->w;
+                    memcpy( kuva + out_off, kuva + in_off, image->pitch );
+                }
+            }
+            else
+            {
+                size_t n_bytes = ((size_t) image->w) * ((size_t) image->h);
+                memcpy( kuva, image->pixels, n_bytes );
+                //printf( "loaded image data from png %s\n", png_name );
```
Comment (veikkos) [inline SRC/PORT_IMAGE.CPP]: Maybe remove commented code while at it?

File: SRC/PORT_IMAGE.H [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#ifndef ULTIMATETAPANKAIKKI_PORT_IMAGE_H
+#define ULTIMATETAPANKAIKKI_PORT_IMAGE_H
+namespace tk_port
+{
+    char *replace_extension_with_png( const char *name );
```
Comment (veikkos) [inline SRC/PORT_IMAGE.H]: Minor nitpick but in other files namespaces aren't indented (same on .src).

File: SRC/PORT_TEXT.CPP [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+                uint8_t *scanline = glyph_data + y * glyph_w;
+                if ( scanline[x] ) empty = false;
+                if ( scanline[x] > 1 && x > right1 ) right1 = x;
+                if ( scanline[ix] > 1 && ix < left1 ) left1 = ix;
+            }
+        }
+        font->glyph_nonempty[glyph_index] = !empty;
+        font->glyph_left_one[glyph_index] = left1;
+        font->glyph_right_one[glyph_index] = right1;
+        /*
```
Comment (veikkos) [inline SRC/PORT_TEXT.CPP]: Maybe remove commented code?

File: SRC/WRITE.CPP [hunk 7]
```diff
@@ -1,196 +1,51 @@
...
-    fnt = fopen( name, "rb" );
-    if ( !fnt ) error( "\nFile '%s' not found\n", name );
-    f_x_size[FONT_NUM] = fgetc( fnt );
-    f_y_size[FONT_NUM] = fgetc( fnt );
-    thefonts[FONT_NUM] = (char *) malloc( f_y_size[FONT_NUM] * f_x_size[FONT_NUM] * 256 );
-    fseek( fnt, 256 * 2, 0 );
-    for ( a = 0; a < f_x_size[FONT_NUM] * f_y_size[FONT_NUM] * 256; a++ )
-        thefonts[FONT_NUM][a] = (char) fgetc( fnt );
-    fclose( fnt );
+void loadfonts( const char *name ) {
```
Comment (veikkos) [inline SRC/WRITE.CPP]: New line before `{` :)

File: SRC/PORT_IMAGE.CPP (hunk 3 above)
Comment (akx) [reply SRC/PORT_IMAGE.CPP]: Well, `error()` is a fatal noreturn, but yeah, good hygiene anyway.

File: SRC/PORT_IMAGE.H (hunk 5 above)
Comment (akx) [reply SRC/PORT_IMAGE.H]: Yeah, blame CLion. 😅

**Comment by akx** [comment]
> Could consider putting new files under still existing SRC/EFP directory as SRC directory has lots of files.

I disagree, since we're getting rid of EFPs as a concept here.

**Comment by akx** [comment]
@veikkos Ready for re-review.

**Comment by akx** [comment]
> Would have renamed `.EFP` -> `.PNG` instead of `.png` like rest of the files. :-)

Sure. Added a commit to do that.

Aside: it makes perfect sense really, but on a case-insensitive file system like my Mac's APFS, one needs to `git rm --cached ...; git add` files for Git to pick up the case changes.

**Pull request merged**